### PR TITLE
[SYSTEMDS-3385] Integration of frontend with monitoring backend

### PR DESCRIPTION
**Pull request by MKehayov**
This change has the initial frontend application written in Angular. It is integrated with the backend so that both applications are communicating with REST requests, with this, coordinators and workers can be managed from the Angular app and real-time statistics can be displayed.

### CODE REVIEW

**Comment by j143** [comment]
Hi @MKehayov thanks a lot for the feature. you can add screenshots here too!

**Comment by MKehayov** [comment]
Sure @j143 here - 

Screenshot from 2022-06-20 22-04-30
Screenshot from 2022-06-20 22-03-43
Screenshot from 2022-06-20 22-02-26
Screenshot from 2022-06-20 22-01-55
Screenshot from 2022-06-20 22-01-39

let me know if you think further changes are required for the initial implementation.

**Comment by j143** [comment]
thanks @MKehayov , I will tryout this feature with some experiments and let you know how it goes.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java [hunk 1]
```diff
@@ -68,13 +83,43 @@ public void setStats(final List<BaseEntityModel> stats) {
...
+	public void setJitCompileTime(final double jitCompileTime) {
+		_jitCompileTime = jitCompileTime;
+	}
+
+	public String getRequestTypeCount() {
+		List<String> requestTypeCountStrArr = new ArrayList<>();
+
+		if (_requestTypeCount != null) {
+			for(var entry : _requestTypeCount) {
+				requestTypeCountStrArr.add(String.format("{" +
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java:99]: this is not nice, make the formatting string outside the for loop (i do not trust the java compiler to fix this.)

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java [hunk 2]
```diff
@@ -68,13 +83,43 @@ public void setStats(final List<BaseEntityModel> stats) {
...
+	}
+
 	@Override
 	public String toString() {
 		return String.format("{" +
 				"\"id\": %d," +
 				"\"name\": \"%s\"," +
 				"\"address\": \"%s\"," +
+				"\"isOnline\": %b," +
+				"\"jitCompileTime\": %.2f," +
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java:120]: same here, make a static string or something different.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java [hunk 3]
```diff
@@ -74,25 +97,33 @@ public double getMemoryUsage() {
...
 				String instruction = entry.getKey();
 				Long count = entry.getValue().getLeft();
 				double duration = entry.getValue().getRight();
-				sb.append(String.format("{" +
-					"\"instruction\": %s," +
-					"\"count\": \"%d\"," +
-					"\"duration\": \"%.2f\"," +
-					"},", instruction, count, duration));
+				heavyHittersStrArr.add(String.format("{" +
+					"\"instruction\": \"%s\"," +
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java:118]: make format string outside loop

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java [hunk 4]
```diff
@@ -103,21 +134,22 @@ public void setHeavyHitterInstructions(final String heavyHitterInstructionsJsonS
...
+		if ((_transferredBytes.isEmpty() || _transferredBytes.isBlank()) && _transferredBytesObj != null) {
+
+			List<String> transferredBytesStrArr = new ArrayList<>();
 
-			sb.append("{");
 			for (var entry: _transferredBytesObj) {
-				sb.append(String.format("{" +
-					"\"datetime\": %s," +
+				transferredBytesStrArr.add(String.format("{" +
+					"\"datetime\": \"%s\"," +
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java:143]: format string outside for loop

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/repositories/DerbyRepository.java [hunk 5]
```diff
@@ -43,17 +43,21 @@ public class DerbyRepository implements IRepository {
 	private static final String ENTITY_SCHEMA_CREATE_STATS_STMT = "CREATE TABLE %s " +
 			"(id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY (START WITH 1, INCREMENT BY 1), " +
 			"%s INTEGER, " +
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/repositories/DerbyRepository.java:45]: this is a good example of pre-computing your format string

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/services/StatsService.java [hunk 6]
```diff
@@ -45,13 +50,20 @@ public static BaseEntityModel getWorkerStatistics(Long id, String address) {
...
+						aggFedStats.coordinatorsTrafficBytes,
+						aggFedStats.requestTypeCount);
 			}
-		} catch(DMLRuntimeException dre) {
+		} catch (DMLRuntimeException dre) {
 			// silently ignore -> caused by offline federated workers
+			log.error("Worker offline: " + dre.getMessage());
 		} catch (Exception e) {
-			throw new RuntimeException(e);
+			log.error("Error: " + e.getMessage());
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/services/StatsService.java:66]: this is super dangerous. 
Please verify that this is the actual behavior you want, 
can you limit the exception caught to something else?
And if it really is the functionality you want add a comment saying that we do not handle the exception but just log it, and that it is fine.

**Review by Baunsgaard — COMMENTED** [review]
This looks great! but we are missing tests.

I don't think we need the file:
Package-lock.json
can this not be git ignored as well?

and tsconfic.spec.json complain about the license not being valid.
In this case just remove it and set it to ignore this file in our pom.xml

best regards
Sebastian

**Comment by j143** [comment]
I have only verified the angular UI, (I will test the full functionality while running the federated experiments shortly.)

1. we can customize README and add license!
2. we can remove package-lock.json. Although, the standard practice is to include in a web repo, here we can keep the code concise by removing the generated code as much as possible.
3. You can add documentation (like a small diagram of the data model if possible), so that people can easily extend it.
4. Of course, as Sebastian mentioned let's add test for each component.
5. Also, let only reveal the enough info in the logs - if need be we can extend it later.

Best,
Janardhan

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java (hunk 1 above)
Comment (MKehayov) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java]: Done, moved logic into a separate file.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java (hunk 2 above)
Comment (MKehayov) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/NodeEntityModel.java]: Same, moved logic into a separate file.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java (hunk 3 above)
Comment (MKehayov) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java]: Done, moved logic into a separate file.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java (hunk 4 above)
Comment (MKehayov) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/models/StatsEntityModel.java]: Done, moved logic into a separate file.

File: src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/services/StatsService.java (hunk 6 above)
Comment (MKehayov) [reply src/main/java/org/apache/sysds/runtime/controlprogram/federated/monitoring/services/StatsService.java]: Completely agree, for now, I left it as it is with additional comments, the problem was that there is a separate thread running for worker stats gathering and if the method fails, the thread stops, and no further data is gathered from any of the workers, the way to fix it is to restart the whole app. With this, the app can continue functioning without the restart.

**Comment by MKehayov** [comment]
Hey, @j143 thank you for the points, my fixes on them are the following:

1. I have added README files for the angular UI and backend, a license is included in the angular one.
2. I agree, that it is a standard practice to have the package-lock in the repo, it is currently there, but let me know if I should remove it.
3. Added in the README file of the backend and the API reference can be found in the angular README.
4. I added at least one test for every component, they currently sum up to 36.
5. I removed the logging of the errors in the StatsService, let me know where else you think we should remove the logs.

Best,
Mito

**Comment by j143** [comment]
> I agree, that it is a standard practice to have the package-lock in the repo, it is currently there, but let me know if I should remove it.

Yes, lets remove it.

**Comment by j143** [comment]
I am facing issues with setting up the federated experiments, could not verify this complete feature at the moment. I will do this as soon as possible.

**Comment by MKehayov** [comment]
Hi @j143, did you manage to run and test the app, and if not, can I help with anything?

**Comment by j143** [comment]
> Hi @j143, did you manage to run and test the app, and if not, can I help with anything?

can you help me set it up. my email is janardhan@apache.org

**Comment by j143** [comment]
Hi @MKehayov - view is not responding/updating when I am trying to add either worker/coordinator.

**Comment by MKehayov** [comment]
Hey @j143 I found a bug, which also prevents the registering of worker/coordinator in the system, and pushed a fix for it, if this does not fix the problem for you, let me know. With the latest changes, I started an implementation of an event timeline for each worker and extended the worker dashboard itself, attached are some screenshots:
Screenshot from 2022-07-29 13-27-58
Screenshot from 2022-07-30 11-00-40

**Comment by j143** [comment]
I tried the latest change set, 😃

1. The worker, coordinator and event timeline looks good.
2. and for `npm run test` I received the following:

image

Thanks for working on this feature.✨

**Comment by j143** [comment]
Looks Good to me. 👍 

Thanks a lot - @MKehayov. ✨🎉

**Pull request closed without merging**